File: DSPractice/queue.cpp
#include <iostream>
#include <stdexcept>
using namespace std;

void error(const string s){
    throw runtime_error(s);
}

class Node{
    public:
        Node* next;
        int val;

        Node(){
            this->next = nullptr;
        }

        Node(const int val){
            this->val = val;
            this->next = nullptr;
        }
};

class Queue{
    private:
        Node *front, *rear;
    public:
        bool isEmpty(){
            return !front && !rear;
        }

        void push(const int val){
            Node* node = new Node(val);
            if(isEmpty()){
                front = rear = node;
                return;
            }
            
            rear->next = node;
            rear = node;
        }

        void pop(){
            Node* node = front;

            if(front == rear){
                front = rear = nullptr;
                delete[] node;
                return;
            }

            front = front->next;
            delete[] node;
        }

        int peekFront(){
            return front->val;
        }

        int back(){
            return rear->val;
        }
};

int main(){
    Queue q;
    q.push(1);
    q.push(2);
    cout<<q.peekFront()<<endl;
    q.pop();
    cout<<q.peekFront()<<endl;
    q.push(5);
    q.push(4);
    q.pop();
    q.pop();
    q.push(9);
    cout<<q.peekFront()<<endl;
    cout<<q.back()<<endl;
    return 0;
}

File: cmd/queue/queue.go
package queue

type Node[T any] struct {
	next *Node[T]
	prev *Node[T]
	val  T
}

type Queue[T any] struct {
	head   *Node[T]
	tail   *Node[T]
	length int
}

func NewQueue[T any]() *Queue[T] {
	return &Queue[T]{}
}

func (q *Queue[T]) Size() int {
	return q.length
}

func (q *Queue[T]) Flatten() []T {
	ls := []T{}

	head := q.head

	for ; head != nil; head = head.next {
		ls = append(ls, head.val)
	}

	return ls
}

func (q *Queue[T]) Push(val T) {
	q.length++

	node := &Node[T]{
		val: val,
	}

	if q.head == nil {
		q.head = node
		q.tail = node
		return
	}

	q.tail.next = node
	node.prev = q.tail
	q.tail = node
}

func (q *Queue[T]) Pop() T {
    q.length--

    if q.length == 0 {
        val := q.head.val

        q.head = nil
        q.tail = nil

        return val
    }

    val := q.head.val

    q.head = q.head.next

    return val
}

func (q *Queue[T]) Front() T {
    return q.head.val
}
